Library

AEscobarCruz opened the pull request:
Major changes which basically sum up to Geospace => (geospace-http, geospace-library), all tests succeed, and code has been updated to use rojoma-json-v3.

urmilan commented:
Remove dependency graph from git.

cloudbees-pull-request-builder commented:
[geospace #213](https://socrata-oss.ci.cloudbees.com/job/geospace/213/) FAILURE
Looks like there's a problem with this pull request

cloudbees-pull-request-builder commented:
[geospace #214](https://socrata-oss.ci.cloudbees.com/job/geospace/214/) FAILURE
Looks like there's a problem with this pull request

velvia commented:
@AEscobarCruz can you describe in more detail where geospace-lib is going to be used, and how?  This is important, as Geospace has high memory requirements that the lifecycle needs to be investigated carefully.

urmilan commented on geospace-http/src/main/scala/com/socrata/geospace/http/client/CoreServerClient.scala:
```diff
@@ -117,7 +117,7 @@ class CoreServerClient(httpClient: HttpClient,
         val request = buildRequest(rb)
         logger.info("Request: " + request)
-        for (response <- httpClient.execute(request)) yield {
+        for (response <- httpClient.execute(request)) {
```
Does this actually work? I would have thought that removing the yield makes T Unit, which means the post method above can't return Try[JValue]. Am I missing something?

rjmac replied:
That's actually the big change in simple-arm v2; `Managed#map` changed from having type `[B](f: A => B): B` to `[B](t: A => B): Managed[B]` (i.e., it's a functorial map, not a way to run the compuation).  Which meant that a new way to run the compuation needed to be introduced; `Managed#run` is the canonical way to do that, but since `(for(...) yield { blah }).run` is hideous and `foreach` is not fundamental to a monadic compuation in the way `map` and `flatMap` are (indeed for most monads `foreach` makes essentially no sense), I made `Managed#foreach` return the value.

AEscobarCruz replied:
Let me investigate, i believe it does work, something about managed not being expected any longer.

urmilan replied:
Oh neat! Thanks for the explanation.

urmilan commented on geospace-http/src/test/scala/com/socrata/geospace/http/GeospaceServletSpec.scala:
```diff
@@ -18,48 +12,22 @@ import org.scalatra.test.scalatest._
-                                                 curatorConfig.maxRetries,
-                                                 RetryOnAllExceptionsDuringInitialRequest,
-                                                 throw ServiceDiscoveryException("No Soda Fountain servers found"))
+class GeospaceServletSpec extends FakeSodaFountain {
```
Yay! Thanks for getting this working.

AEscobarCruz replied:
:)

urmilan commented:
The test shapefiles under /data shouldn't need to be copied into both projects. They are only used by ShapefileReaderSpec.

AEscobarCruz commented:
We will be using delta-importer-2 (di2) as our ingress route, as such, we will need a GeoCollector service (in charge of importing files from client and pushing jobs to di2 ) and a GeoWorker service (takes jobs from queue, pulls files, and does the transformations for ingress). Therefore GeoWorker needs a geospace-library, in order to avoid duplication of code, import  code as a library, and finally not pull unnecessary service definitions.

urmilan commented:
@velvia From what I understand Alejandro needs to use the shapefile ingestion code in delta importer. I don't think he is using any of the spatial caching logic that we have memory issues with (@AEscobarCruz, correct me if I'm wrong).

cloudbees-pull-request-builder commented:
[geospace #216](https://socrata-oss.ci.cloudbees.com/job/geospace/216/) FAILURE
Looks like there's a problem with this pull request

AEscobarCruz commented:
@urmilan ShapefileReaderSpec and GeospaceServletSpec both use the data dir, they may not share both sets of data, so I could split it up.

urmilan commented on geospace-lib/src/main/scala/com/socrata/geospace/lib/regioncache/RegionCache.scala:
```diff
@@ -104,10 +110,16 @@ abstract class RegionCache[T](maxEntries: Int = 100) //scalastyle:ignore
         regionIndexLoadTimer.time {
           payload.toOption.
-            flatMap { jvalue => GeoJson.codec.decode(jvalue) }.
+            flatMap {  jvalue => GeoJson.codec.decode(jvalue) match {
```
```scala
GeoJson.codec.decode(jvalue).right.toOption
```
?

AEscobarCruz replied:
yup!

urmilan commented on geospace-lib/src/main/scala/com/socrata/geospace/lib/regioncache/SpatialIndex.scala:
```diff
@@ -69,7 +71,7 @@ object SpatialIndex {
    * Create a SpatialIndex[String] from a Layer/FeatureSource.  The feature ID will be stored in the index.
    *
-   * @param layer an [[org.geoscript.layer.Layer]] or GeoTools FeatureSource.
+   * @param layer an [org.geoscript.layer.Layer] or GeoTools FeatureSource.
```
Why? See Scaladoc style guide on referencing classes: http://docs.scala-lang.org/style/scaladoc.html

urmilan commented on geospace-lib/src/main/scala/com/socrata/geospace/lib/regioncache/SpatialIndex.scala:
```diff
@@ -1,7 +1,9 @@
+import com.socrata.geospace.lib.Utils
+import com.socrata.geospace.lib.feature.FeatureExtensions
+import FeatureExtensions._
+import Utils._
```
What's the reason for breaking this up into 2 imports?

velvia replied:
By the way, it's better to not do relative imports like `FeatureExtensions._`, but include the full path instead for clarity.

urmilan commented on geospace-lib/src/main/scala/com/socrata/geospace/lib/utils/GeoServeletUtils.scala:
```diff
@@ -0,0 +1,10 @@
+
+/**
+ */
+object GeoServeletUtils {
```
Remove this empty class if it isn't being used anywhere.

urmilan commented on geospace-lib/src/test/scala/com/socrata/geospace/lib/client/SodaResponseSpec.scala:
```diff
@@ -1,10 +1,9 @@
-import com.rojoma.json.ast.{JString, JObject}
-import com.socrata.geospace.client.SodaResponse._
+import com.rojoma.json.v3.ast.{JObject, JString}
+import com.socrata.geospace.lib.client.SodaResponse.{JsonParseException, _}
```
Would ```import com.socrata.geospace.lib.client.SodaResponse._``` suffice?

urmilan commented on project/BuildSettings.scala:
```diff
@@ -0,0 +1,64 @@
+      scalaVersion := "2.10.3",
+      organization := Organization,
+      autoAPIMappings := true,
+      apiMappings += (
```
What are these "example.org" autoAPIMappings and apiMappings settings used for? If we're not actually using them for anything real we should remove them.

AEscobarCruz replied:
actually, it was my attempt to get
@param layer an [[org.geoscript.layer.Layer]] or GeoTools FeatureSource.

to work, as it will not compile, in a previous pull request, I tried to address this, and was given some suggestions on how about doing it. It hasn't worked, so this is still an open item to address.

urmilan commented on project/CommonDependencies.scala:
```diff
@@ -0,0 +1,38 @@
+  private val socrataDeps = Seq(
+    "com.rojoma"               %% "rojoma-json-v3"         % "3.2.1",
+    "com.rojoma"               %% "simple-arm-v2"          % "[2.0.0,3.0.0)",
+    "com.socrata"              %% "socrata-http-client" % "3.0.0",
```
Nit; can you line these back up? :)

urmilan commented on project/build.scala:
```diff
@@ -1,147 +1,20 @@
-  )
+  // originally had aggregate(allOtherProjects: _*) but gave an error
+  // [error] java.lang.IllegalAccessException: Class Build$$anonfun$allOtherProjects$2 can not access a member of class Build$ with modifiers "private"
+  // adding projects directly helps, but to determine cause.
```
If that pattern never existed in the code for you to remove, I don't think you need to include this comment.

velvia commented:
@AEscobarCruz @urmilan actually the Shapefile ingress also uses a lot of memory.  It fails on very large shapefiles because it does the reprojection in memory.  If you look in the code there should be TODOs about making it write to disk or streaming.  That could be one potential reason for memory issues.

urmilan commented on version.sbt:
```diff
@@ -1 +1 @@
-version in ThisBuild := "0.3.5-SNAPSHOT"
\ No newline at end of file
+version in ThisBuild := "0.3.6-SNAPSHOT"
```
I _think_ we probably need to increment more than just the patch version if we've moved to new major versions of many of the dependencies, but I'm not sure what the best practice is... @velvia/@rjmac?

velvia replied:
We should definitely bump the minor version at the very least.  0.4.0.

urmilan commented:
@velvia Hmm, that's true. @AEscobarCruz can you look into whether Geotools supports streaming in the shapefile rather than loading it fully into memory as we do now?

AEscobarCruz commented:
@velvia @urmilan I think we have to consider how a worker object performs a task also. I will try to investigate the issue as we progress along and report any findings as I go along. Sounds like we should build in some sort of stress test.

velvia commented:
@AEscobarCruz stress test is definitely a good idea.

@urmilan what you would do is, instead of generating a standard Scala collection of features during reprojection, you use GeoTools to write the output to a new Shapefile (or other on disk format), and you use an Iterator so only a few features need to be in memory at one time.  Or better yet, just reproject when we are transmitting the data to core.

cloudbees-pull-request-builder commented:
[geospace #217](https://socrata-oss.ci.cloudbees.com/job/geospace/217/) FAILURE
Looks like there's a problem with this pull request

cloudbees-pull-request-builder commented:
[geospace #219](https://socrata-oss.ci.cloudbees.com/job/geospace/219/) FAILURE
Looks like there's a problem with this pull request

AEscobarCruz commented:
@velvia  Correct me if I am wrong, but what you are saying is that during ingress, the files are transformed all in memory, instead of caching changes locally and then sending the deltas along when completed?

velvia commented:
@AEscobarCruz not quite.  It's not about deltas, because Shapefiles are always processed in their entirety (it's a .zip format, you have to unpack and process everything).  There is no caching involved, Geospace has to send everything to Core.

The steps are:
1) Read features from unpacked shapefile .dbf
2) newFeatures = features.map(reproject _)
3) Write newFeatures to Core

It's about trying to get steps 1, 2, and 3 to be done all incrementally, so that only a few features (ideally 1) need to be kept in memory at once, rather than doing them in separate steps such that intermediate results are all in memory.

urmilan commented:
LGTM. Please bump the minor or major version as I mentioned in an earlier comment, before you merge.

AEscobarCruz commented:
Squashed to 1 commit, updated the minor version. Guys I would like you guys to build it yourselfs and see if you have any concerns. Each project builds and compiles for me just fine, so I dont think we have any issues. I would also like to update the onramp scrips because I think they will be broken after the merge. ALSO jenkin build scripts will most likely need updates, but I dont forsee that being too difficult to deal with?

cloudbees-pull-request-builder commented:
[geospace #233](https://socrata-oss.ci.cloudbees.com/job/geospace/233/) FAILURE
Looks like there's a problem with this pull request

cloudbees-pull-request-builder commented:
[geospace #234](https://socrata-oss.ci.cloudbees.com/job/geospace/234/) FAILURE
Looks like there's a problem with this pull request

sokrahta commented:
don't forget to address the cloudbees merge failure!

cloudbees-pull-request-builder commented:
[geospace #238](https://socrata-oss.ci.cloudbees.com/job/geospace/238/) FAILURE
Looks like there's a problem with this pull request

cloudbees-pull-request-builder commented:
[geospace #240](https://socrata-oss.ci.cloudbees.com/job/geospace/240/) FAILURE
Looks like there's a problem with this pull request

cloudbees-pull-request-builder commented:
[geospace #254](https://socrata-oss.ci.cloudbees.com/job/geospace/254/) FAILURE
Looks like there's a problem with this pull request

cloudbees-pull-request-builder commented:
[geospace #255](https://socrata-oss.ci.cloudbees.com/job/geospace/255/) FAILURE
Looks like there's a problem with this pull request

cloudbees-pull-request-builder commented:
[geospace #257](https://socrata-oss.ci.cloudbees.com/job/geospace/257/) FAILURE
Looks like there's a problem with this pull request

cloudbees-pull-request-builder commented:
[geospace #259](https://socrata-oss.ci.cloudbees.com/job/geospace/259/) SUCCESS
This pull request looks good

urmilan commented:
:+1: Thanks for the epic refactor @AEscobarCruz. Merging now...
